mozilla/fx-private-relay
# MPP-3439: Refactor DataIssueTask / CleanerTask, clean users with blank emails

Some users in production have `user.email == ""`. This prevents us from forwarding them any emails.

At least one also has an email in their Mozilla account data in the database. The `user.email` should have been set to their `socialaccount.extra_data["email"]` during signup, but some other issue must have prevented it from happening, or caused it to be cleared.

This PR adds a cleaner `MissingEmailCleaner` to the cleaner lists, so that this issue can be fixed in production. The cleaners are run daily as a Cron Job in production and other environments, so these and future incidents will be fixed.

# The Refactor

The changes for MPP-3439 are small compared to the refactor of `DataIssueTask` and the derived `CleanerTask`. However, none of the tests for the existing email cleaners changed, so the refactor (hopefully!) works the same.

It is probably more enlightening to read the code in `emails/cleaners.py` and `privaterelay/cleaners.py` than to try to understand all the code changes, or the explanation that follows. The purpose of all the refactoring is to make these easier to read and write, and require less repetitive code. Feel free to skip ahead to "How to Test".

Previously a cleaner would derive from `CleanerTask` and define the methods:

* `_get_counts_and_data` - Make database `COUNT` queries, and create a two-level dictionary of counts, a "summary" plus one for each model. Return the counts and querysets for getting the problem rows.
* `_clean` - Take the "problem row" querysets, and update the rows or iterate over the rows and fix them one at a time. Return the count of fixed rows.
* `markdown_report` - Generate a blob of text representing the data and how many rows were cleaned.

After the refactor, these functions have been generalized and moved up into the parent classes. Because there is quite a lot of shared code, it has been moved to `privaterelay/cleaner_task.py`, leaving `privaterelay/cleaners.py` for the new `MissingEmailCleaner`.

After the refactor, a data cleaner derives from `CleanerTask` and defines a `data_specification` with two models:

* `DataModelSpec` - This represents a model, such a `User` or `RelayAddress`, as well as some details for data cleaning
* `DataBisectSpec` - A filter of a model or a further filtering, as well as the opposite of that filter. For example, `User.objects.filter(is_active=True)` and `User.objects.exclude(is_active=True)` bisects the `User` table into two sets of rows. All the current cleaning operations can be represented as a series of filters that bisect the rows until the problem rows, if any, remain.

The data cleaner class also defines functions with the pattern `clean_<plural model_name>`. These are passed a queryset representing the problem rows, if any. The function fixes the issues if it can, and returns a count of the issues it fixed.

The `CleanerTask` uses some new classes (`ReportItem`, `CleanedItem`, `BaseDataItem`, `DataModelItem`, and `DataItem`) to implement querying, generating counts, running cleaning (in cooperation with the `clean_x` methods) and creating reports. These mostly hold and validate data, and are thoroughly tested.

The unused `DetectorTask` is gone. The base class `DataIssueTask` could do the same, or be extended if we ever want to use this framework to explore the data but not fix it.

Other changes:

* Added `verbose_name_plural` for `RelayAddress` and `DomainAddress`. This changes them from "addresss" to "addresses" in the Django admin, and also means it can be used in `DataTask` where a dictionary key is needed.
* Added `"raise NotImplementedError"` to the coverage ignore list, so `@abstractmethod` functions do not appear as uncovered.

# How to test:

In your local environment, ensure you have at least one staff user (can access the Django admin) and one regular user (logs in with Mozilla Accounts).

As the staff user, go to http://127.0.0.1:8000/admin/auth/user/
  - From the user list, select your regular user.
  - From the user detail page, clear the email then click the "Save" button.
  - On the user list, confirm the regular user's email is clear.

On this branch, run:
```
./manage.py cleanup_data -v 2 --missing-email --clean
```

The output contains a log entry and a text report, and is something like:
```
{"Timestamp": 1716414385805936896, "Type": "eventsinfo.cleanup_data", "Logger": "fx-private-relay", "Hostname": "jwhitlock-MB-M2.lan", "EnvVersion": "2.0", "Severity": 6, "Pid": 17036, "Fields": {"cleaned": true, "timers": {"query_s": 0.002, "clean_s": 0.002}, "tasks": {"missing-email": {"counts": {"summary": {"ok": 5, "needs_cleaning": 1, "cleaned": 1}, "users": {"all": 7, "email": 5, "!email": 2, "!email.fxa": 1, "cleaned": 1}}, "timers": {"query_s": 0.002, "clean_s": 0.002}}}, "rid": null, "msg": "cleanup_data complete, cleaned 1 of 1 issue."}}

# Summary

|    task     |issues|fixed|query (s)|fix (s)|
|------------:|-----:|----:|--------:|------:|
|missing-email|     1|    1|    0.002|  0.002|
|      _Total_|     1|    1|    0.002|  0.002|

# Details
## missing-email
When User.email is empty, we are unable to forward email to the Relay user. We
can get the email from the FxA profile if available.

Detected 1 issue in 0.002 seconds.
Cleaned 1 issue in 0.002 seconds.

Users:
  All: 7
    Email   : 5 ( 71.4%)
    No Email: 2 ( 28.6%)
      Has FxA: 1 ( 50.0%)
        Cleaned: 1 (100.0%)

```

As the staff user, refresh the user list at http://127.0.0.1:8000/admin/auth/user/ . The regular user should have their email again.

As a bonus, run all the cleaners:

```
./manage.py cleanup_data -v 2 --clean
```
